app/repositories/SessionRepository.scala

```diff
@@ -130,57 +128,109 @@ class DefaultSessionRepository @Inject() (config: AppConfig,
+    findAndUpdate(
+      Json.obj("_id" -> key),
+      Json.obj(
+        (List(toFieldPair("$set", Json.obj(toFieldPair(TtlExpiryFieldName, Json.obj("$date" -> Instant.now().toEpochMilli))))) ++
+              pageUrl.fold[List[FieldAttr]](Nil)(url => List("$push" -> Json.obj(PageHistoryKey -> PageHistory(url, Nil))))
+        ).toArray: _*),
+      fetchNewObject = false
+    ).flatMap { r =>
+      r.result[DefaultSessionRepository.SessionProcess]
+      .fold {
+        logger.warn(s"Attempt to retrieve cached process from session repo with _id=$key returned no result, lastError ${r.lastError}")
+        Future.successful(Left(NotFoundError): RequestOutcome[ProcessContext])
+      }{ sp => // SessionProcess returned by findAndUpdate() is intentionally that prior to the update!!
+        pageUrl.fold[Future[RequestOutcome[ProcessContext]]](
+          Future.successful(Right(ProcessContext(sp.process, sp.answers, sp.labels, sp.flowStack, sp.continuationPool, sp.pageMap, Nil, None)))
+        ){url =>
+          sp.pageMap.get(url.drop(sp.process.meta.processCode.length)).fold[Future[RequestOutcome[ProcessContext]]]{
+            logger.warn(s"Attempt to move to unknown page $url")
+            Future.successful(Left(NotFoundError))
+          }{pageDesc =>
+            logger.debug(s"Incoming Page: ${pageDesc.id}, $url, current legalPageIds: ${sp.legalPageIds}")
+            if (sp.legalPageIds.isEmpty || sp.legalPageIds.contains(pageDesc.id)){ // Wild card or fixed list of valid page ids
+              val firstPageUrl: String = s"${sp.process.meta.processCode}${sp.process.startUrl.getOrElse("")}"
+              val (backLink, historyUpdate, flowStackUpdate, labelUpdates) = sessionProcessTransition(url, sp, previousPageByLink, firstPageUrl)
+              val labels: Map[String, Label] = sp.labels ++ labelUpdates.map(l => (l.name -> l)).toMap
+              val legalPageIds = Process.StartStanzaId :: pageDesc.id :: pageDesc.next ++
+                                  backLink.fold[List[String]](Nil)(bl => List(sp.pageMap(bl.drop(sp.process.meta.processCode.length)).id)).distinct
+              val processContext = ProcessContext(
+                                    sp.process,
+                                    sp.answers,
+                                    labels,
+                                    flowStackUpdate.getOrElse(sp.flowStack),
+                                    sp.continuationPool,
+                                    sp.pageMap,
+                                    legalPageIds,
+                                    backLink)
+              logger.debug(s"NewLegalIds: ${legalPageIds}")
+              saveUpdates(key, historyUpdate, flowStackUpdate, labelUpdates, legalPageIds).map {
                 case Left(err) =>
                   logger.error(s"Unable to save backlink history, error = $err")
                   Right(processContext)
                 case _ => Right(processContext)
```

Comment:
Should we return the processContext here or an error. If something goes wrong with a flow update or label updates should we continue?

Reply:
done